### ISSUE TITLE
Add support for arbitrary parameters when binding services via manifest.yml

### ISSUE THREAD

**Issue by geoffwa** [cloudfoundry/cli#1173]
The CLI doesn't currently support arbitrary parameters for binding services via a manifest.yml file: https://docs.cloudfoundry.org/devguide/services/application-binding.html#-binding-with-application-manifest

For NFS volume mounts: https://docs.cloudfoundry.org/devguide/services/using-vol-services.html#create it's quite useful to specify the mount path and uid/gids when binding an NFS service to one or more apps.

Are there any blockers to adding this support?

**Comment by dkoper** [comment]
Hi @geoffwa 

I've been exploring this in this story: https://www.pivotaltracker.com/story/show/117493797.
Progressing slowly as other high-priority things came up, and the refactor of `push` and `create-app-manifest` (pre-requisites to this work) have not neared completion yet.
Also, the current manifest schema doesn't seem to allow us to add support for this easily (see story), and the additional complexities of adding support for route services.

Regards,
Dies Koper
CF CLI PM

**Comment by felixfriedrich** [comment]
Just for the record, I'd really like to see this feature. We're using the gcp-service-broker for Google storage buckets. The bind command for these buckets expects the role of the user as a parameter. Not being able to put this into the manifest is quite a problem at the moment :-)

Thanks!

Felix

**Comment by felixfriedrich** [comment]
@dkoper Do I understand you correctly that fixing this is expected to take some time?

**Comment by dkoper** [comment]
@felixfriedrich Correct.

**Comment by JohannesRudolph** [comment]
I'd also like to see this feature. It's essential when working with volume services and very useful for other services as well.

**Comment by tnwang** [comment]
This would be a great feature to support via app manifest. The value of this feature shines when there are per-service-binding configurations that are important, not just per-service-instance configurations.

We have a service broker today that reads from the app manifest for separate variables, with custom logic to understand this. However, we also plan to support service bindings parameters which means it will take a different set of values elsewhere given that the two flows cannot be the same today.

Having the ability to do this through one channel allows for much more maintainable service-broker code if we want to provide the streamlined service broker integration experience for people pushing apps on CF.

**Comment by Quinity** [comment]
I agree, this should be implemented and would provide new possibilities for providing services when deploying applications and binding services.

**Comment by eruvanos** [comment]
Any update on this?

**Comment by mulderga-cgi** [comment]
Having the same problem here with this NFS service in our deployment pipeline ... it's not an option for us to push with `--no-start` and add the binding manually. The only option is to make support arbitrary parameters when binding services via manifest.yml

**Comment by georgi-lozev** [comment]
It's the very first thing in the list, I've started the doc with it.

**Comment by jeroenmies** [comment]
Any idea when this might be finished? We have the same issue here and it is preventing us from upgrading our infrastructure and getting working access to the file system in the company?

**Comment by williammartin** [comment]
@jeroenmies Is your problem also regarding the NFS broker? If not can you tell us a little bit more about the problem.

@mattmcneeney Would we consider prioritising this before v7, there seems to be quite a few people desiring it?

**Comment by jeroenmies** [comment]
Yes it is the same issue.
When deploying a new version of our app through our pipeline deployment (using manifest files) the new instance needs to bind against the existing NFS service, but the manifest does not allow for specifying the binding information, hence we lose the connection to NFS.

Binding every instance manually obviously is not an option.

**Comment by mattmcneeney** [comment]
@abbyachau I'm wondering how often you've heard this request before? The SAPI team may be able to work on this alongside the transition to v7, however, I know we are hesitant to add too much to v6.

@tjvman Do you have a sense of how much work it would be to support declaring parameters for a _synchronous_ service binding in the manifest?

**Comment by abbyachau** [comment]
cc @zrob 

hey @geoffwa etc is there a current workaround at moment for this workflow? 

hey @mattmcneeney I've not heard this often but based on the history of this issue, looks like there's a good amount of support. I'd like to gather a bit more information first, if there's an existing workaround, for example.

**Comment by Samze** [comment]
@mattmcneeney @abbyachau @tjvman If we want to do this in v7 then we would add it to the [Server side manifest](http://v3-apidocs.cloudfoundry.org/version/3.59.0/index.html#the-app-manifest-specification). So the changes (and all services manifest requests) would mainly be in the API rather than the CLI.

**Comment by abbyachau** [comment]
@Samze thanks, absolutely agree this ideally and preferably should be in v7, and on v3 server side manifests.

**Comment by henryaj** [comment]
@zrob - any thoughts on [Sam's comment](https://github.com/cloudfoundry/cli/issues/1173#issuecomment-427018547)?

**Comment by pvtl-dnoguerol** [comment]
Another vote for this on behalf of a PAS customer. It would further facilitate consolidated declarative application deployment configuration.

**Comment by zrob** [comment]
We're definitely tracking this for server side manifest https://www.pivotaltracker.com/story/show/158669166

We expect to pick it up once the cli has adopted the new manifest style.

**Comment by mendarie** [comment]
we are really needing this for all of the GCP connectors!!  all of the GCP bindings require a roll to be passed!!!

https://docs.pivotal.io/partners/gcp-sb/using.html

**Comment by wenxin-liu** [comment]
We're also having issues without this feature being implemented, for using the gcp-service-broker, as role needs to be passed as a parameter upon binding service instance.

**Comment by abbyachau** [comment]
cc @tcdowney @ssisil (CC API anchor and PM)

**Comment by asaikali** [comment]
any updates on a timeline when this issue will be resolved?

**Comment by tcdowney** [comment]
@asaikali Sorry, I just now noticed the tag from Abby. As Zach pointed out above, we have a placeholder story [#158669166](https://www.pivotaltracker.com/n/projects/966314/stories/158669166), but it's empty and hasn't been scheduled.

Tagging our (CAPI) PM @ssisil again to make sure he's aware of this request.

Currently the [server-side app manifests](https://v3-apidocs.cloudfoundry.org/version/3.68.0/index.html#app-manifest) support the ability to create (but not update) service bindings via the following syntax.

```yaml
applications:
- name: my-app
  services:
  - my-service1
  - my-service2
```

We'd have to update this to be more than just an array of strings, but it seems like we could do something like this:
```yaml
applications:
- name: my-app
  services:
  - name: my-service1
    parameters: { }
  - name: my-service2
    parameters: { }
```

**Comment by mulderga-cgi** [comment]
@tcdowney that would be a great feature to have! 👍 

I would suggest that you don't use the curly braces ... key-value pair(s) would be more interesting:

```yaml
--- 
applications: 
- name: my-app
  services:
  - name: my-service1
    parameters: 
      foo: bar
      herp: derp
  - name: my-service2
    parameters: 
      foo: bar
```

**Comment by eruvanos** [comment]
> ```yaml
> foo: bar
>       herp: derp
> ```

Hey, curly braces and key-value pairs are the same in YAML, just different style. 

Check: https://en.wikipedia.org/wiki/YAML

**Comment by tcdowney** [comment]
We just added support for this to the server-side app manifest in commit https://github.com/cloudfoundry/cloud_controller_ng/commit/c3437fd27933a1cd085f721e96f92bc6e0827f5b (story: [#158669166](https://www.pivotaltracker.com/story/show/158669166)) and it should ship in the next capi-release (`v1.80.0`). So in other words, coming soon!

See: http://v3-apidocs.cloudfoundry.org/version/release-candidate/#app-manifest

cc @abbyachau @acrmp in case you're tracking this on the CLI side.

**Comment by abbyachau** [comment]
Hi folks, I'm moving this over the CAPI github repo since they will be handling server side manifest moving forward in V3. Thanks.
